File: src/MyServiceBus.AzureServiceBus/AzureServiceBusMessageEntityNameFormatter.cs
using System.Collections.Concurrent;
using System.Reflection;
using System.Text;

namespace MyServiceBus;

/// <summary>
/// Formats message entity names using the MassTransit Azure Service Bus convention.
/// </summary>
public sealed class AzureServiceBusMessageEntityNameFormatter : IMessageEntityNameFormatter
{
    private readonly ConcurrentDictionary<Type, string> _cache = new();

    public static AzureServiceBusMessageEntityNameFormatter Instance { get; } = new();

    private AzureServiceBusMessageEntityNameFormatter()
    {
    }

    public string FormatEntityName(Type messageType)
    {
        ArgumentNullException.ThrowIfNull(messageType);
        var attribute = messageType.GetCustomAttribute<EntityNameAttribute>();
        return attribute?.EntityName ?? _cache.GetOrAdd(messageType, Format);
    }

    private static string Format(Type messageType)
    {
        if (messageType.IsGenericTypeDefinition)
            throw new ArgumentException("An open generic type cannot be used as a message name.", nameof(messageType));

        return Append(new StringBuilder(), messageType, null).ToString().Replace("[]", "__", StringComparison.Ordinal);
    }

    private static StringBuilder Append(StringBuilder builder, Type messageType, string? scope)
    {
        if (messageType.IsGenericParameter)
            return builder;

        var messageNamespace = messageType.Namespace;
        if (messageNamespace is not null && !string.Equals(messageNamespace, scope, StringComparison.Ordinal))
            builder.Append(messageNamespace).Append('/');

        if (messageType.IsNested)
            Append(builder, messageType.DeclaringType!, messageNamespace).Append('-');

        if (!messageType.IsGenericType)
            return builder.Append(messageType.Name);

        var name = messageType.GetGenericTypeDefinition().Name;
        var arityIndex = name.IndexOf('`', StringComparison.Ordinal);
        builder.Append(arityIndex > 0 ? name[..arityIndex] : name).Append("--");

        var arguments = messageType.GetGenericArguments();
        for (var index = 0; index < arguments.Length; index++)
        {
            if (index > 0)
                builder.Append("---");
            Append(builder, arguments[index], messageNamespace);
        }

        return builder.Append("--");
    }
}

File: src/Java/myservicebus-azure-service-bus/src/main/java/com/myservicebus/azure/servicebus/AzureServiceBusMessageEntityNameFormatter.java
package com.myservicebus.azure.servicebus;

import com.myservicebus.EntityName;
import com.myservicebus.MessageEntityNameFormatter;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/** Formats message entity names using the MassTransit Azure Service Bus convention. */
public final class AzureServiceBusMessageEntityNameFormatter implements MessageEntityNameFormatter {
    public static final AzureServiceBusMessageEntityNameFormatter INSTANCE =
            new AzureServiceBusMessageEntityNameFormatter();

    private final Map<Class<?>, String> cache = new ConcurrentHashMap<>();

    private AzureServiceBusMessageEntityNameFormatter() { }

    @Override
    public String formatEntityName(Class<?> messageType) {
        if (messageType == null) {
            throw new IllegalArgumentException("Message type cannot be null");
        }
        EntityName annotation = messageType.getAnnotation(EntityName.class);
        return annotation != null
                ? annotation.value()
                : cache.computeIfAbsent(messageType, AzureServiceBusMessageEntityNameFormatter::format);
    }

    private static String format(Class<?> messageType) {
        if (messageType.isArray()) {
            return format(messageType.getComponentType()) + "__";
        }

        StringBuilder builder = new StringBuilder();
        Package messagePackage = messageType.getPackage();
        if (messagePackage != null && !messagePackage.getName().isBlank()) {
            builder.append(messagePackage.getName()).append('/');
        }
        appendTypeName(builder, messageType);
        return builder.toString();
    }

    private static void appendTypeName(StringBuilder builder, Class<?> messageType) {
        Class<?> enclosingType = messageType.getEnclosingClass();
        if (enclosingType != null) {
            appendTypeName(builder, enclosingType);
            builder.append('-');
        }
        builder.append(messageType.getSimpleName());
    }
}
